stocks-portfolio example (#1735)
opsdroid/opsdroid: docs/examples/stocks-portfolio.md

This example will use [YFinace](https://github.com/ranaroussi/yfinance) to get the information of the stocks and [SQLite3](https://www.sqlite.org/index.html) to keep our portfolio when we close the bot.

*If you need help or if you are unsure about something join our* [matrix channel](https://riot.im/app/#/room/#opsdroid-general:matrix.org) *and ask away! We are more than happy to help you.*

## Building the Skill

We are now ready to commence building our skill! First, you will need to create a folder for the stocks portfolio skill. Choose a location and name it stocks-portfolio.

```bash
mkdir /path/to/my/stocks-portfolio
```

Inside of the directory we will be making a `configuration.yaml` file which we will be using in the next section to config our skill.

### Configuration

Now, let's open `configuration.yaml` file and add our stocks portfolio skill to the skills section.

```yaml
skills:
  stock:
    path: /path/to/my/stocks-portfolio
```

### Imports and Classes Needed

Now that our skill has been configured in the `configuration.yaml` file, we will continue by creating the `__init__.py` file inside of our stocks-portfolio folder and start working on the skill!

> The first thing we need to do is to import the `Skill` class and the `regex_matcher` from opsdroid. Then we will need to import `yfinance` and `sqlite3` modules. Your `__init__.py` file should look like this:

Reviewer: ```suggestion
The first thing we need to do is to import the `Skill` class and the `regex_matcher` from opsdroid. Then we will need to import `yfinance`, we will also use [opsdroid's memory](https://docs.opsdroid.dev/en/stable/skills/memory.html) to save data into the SQL database.. Your `__init__.py` file should look like this:
```

Author: Hello Fabio, I will try to understand the memory thing more before I rewrite the docs. I think maybe a week.

```python
import yfinance as yf

> import sqlite3

Reviewer: ```suggestion

```

from opsdroid.matchers import match_regex
from opsdroid.skill import Skill
```

### Connecting to SQLite3 

Our next step is to connect our `__init__.py` file to a SQLite database. Your `__init__.py` file should look something like this now:

```python
import yfinance as yf
import sqlite3

from opsdroid.matchers import match_regex
from opsdroid.skill import Skill

> conn = sqlite3.connect('/path/to/my/database')

Reviewer: Feel free to ignore the database suggestions, if you want to use the opsdroid memory you can use `self.opsdroid.memory.put(<key>, <object>)
```suggestion

```

c = conn.cursor()
```

> ### Recieving Input

Reviewer: ```suggestion
### Receiving Input
```

> Now we have to collect what the bot will revieve, we will be doing this by first creating a class that inherits the Skill class. Inside we will be using the special opsdroid function which is `@match_regex`. This function will send the message, and to collect the input we will be using `(.*)` beside the message, like this `@match_regex(r"Stock: (.*)")`. We will then be accessing it with `message.regex.group(1)`. We can now assign what we recieve from the bot. 

Reviewer: ```suggestion
Now we have to collect what the bot will receive, we will be doing this by first creating a class that inherits the Skill class. Inside we will be using the special opsdroid function which is `@match_regex`. This function will send the message, and to collect the input we will be using `(.*)` beside the message, like this `@match_regex(r"Stock: (.*)")`. We will then be accessing it with `message.regex.group(1)`. We can now assign what we receive from the bot. 
```

The `StockSkill` class should look something like this:

```python
class StockSkill(Skill):

    @match_regex(r"Stock: (.*)")
    async def search_info_stock(self, message):
        stock_name = message.regex.group(1)
```

### Getting the Data

Now the fun part! We will be using the YFinance module to get the data for the stocks, the module will take the stock's symbol and return the info. You can read the [documentation](https://github.com/ranaroussi/yfinance) for more info. 

Make your `StockSkill` class look something like this:

```python
class StockSkill(Skill):

    @match_regex(r"Stock: (.*)")
    async def search_info_stock(self, message):
        stock_name = message.regex.group(1)
        stock = yf.Ticker(stock_name) 
```

> Now `stock` is what we recieve from YFinance when we input a stock's symbol.

Reviewer: ```suggestion
Now `stock` is what we receive from YFinance when we input a stock's symbol.
```

### Using the Data

> Now after getting the data and assigning it to a variable we can start using it! We can now return the info to the connector. We will be using the info property but if you are interested on the different properties you can use their [documentation](https://github.com/ranaroussi/yfinance.

Reviewer: ```suggestion
After getting the data and assigning it to a variable, we can start using it! We can now return the info to the connector. We will be using the info property, but if you are interested in the different properties, you can use their [documentation](https://github.com/ranaroussi/yfinance).
```

   "heldPercentInsiders":0.16246,
   "nextFiscalYearEnd":1640908800,
   "mostRecentQuarter":1601424000,
   "shortRatio":3.2,
   "sharesShortPreviousMonthDate":1607990400,
   "floatShares":120820703,
   "enterpriseValue":38934634496,
   "threeYearAverageReturn":"None",
   "lastSplitDate":"None",
   "lastSplitFactor":"None",
   "legalType":"None",
   "lastDividendDate":"None",
   "morningStarOverallRating":"None",
   "earningsQuarterlyGrowth":"None",
   "dateShortInterest":1610668800,
   "pegRatio":-15.31,
   "lastCapGain":"None",
   "shortPercentOfFloat":0.048600003,
   "sharesShortPriorMonth":5086368,
   "impliedSharesOutstanding":"None",
   "category":"None",
   "fiveYearAverageReturn":"None",
   "regularMarketPrice":152.33,
   "logo_url":"https://logo.clearbit.com/unity.com"
}
```

We want the `['longName']`, `['regularMarketOpen']` and `['longBusinessSummary']`. Now we want the bot to send the data, so we will be using `await message.respond()`. Here is what it should look like:

```python
class StockSkill(Skill):

    @match_regex(r"Stock: (.*)")
    async def search_info_stock(self, message):
        stock_name = message.regex.group(1)
        stock = yf.Ticker(stock_name)

>         await message.respond(f"Name: {stock.info['longName']}")

Reviewer: I'm happy to have it send more than one message, just as reference you could send a single message and make the string look like:

```python
await message.respond(f"Name: {stock.info['longName']}\nCurrent Price: {stock.info['regularMarketOpen']} {stock.info['currency']}\nAbout: {stock.info['longBusinessSummary']}")
```

Author: Thank you, I will be adding this

        await message.respond(f"Current Price: {stock.info['regularMarketOpen']} {stock.info['currency']}")
        await message.respond(f"About: {stock.info['longBusinessSummary']}")
```

Congratulations! You have now made a bot that can tell you the name, price and info on a stock. You can do this with `Stock: <symbol-of-stock>`. With the basic part of the bot finished we can make the portfolio/saving section.

### Saving and Adding to the Portfolio

> #### Creating the Table

Reviewer: Like I said before, i'm happy to have the SQL steps here -  if you want to replace it you could change this step like:

### Putting in opsdroid memory

Opsdroid has a memory class which can be used to persist data, to use opsdroid memory you need to add the following to your configuration:

```yaml
databases:
  sqlite:
```

Then we can add to the database using `await self.opsdroid.memory.put(stocks, params)`

---

You might need to think about how the data is going to be saved/returned

Author: Thank you this info really helps!

        rows = c.fetchall()

        for row in rows:
            stock = yf.Ticker(row[0])
            await message.respond(f"Name: {stock.info['longName']}")
            await message.respond(f"Price: {stock.info['regularMarketOpen']} {stock.info['currency']}")
            await message.respond(f"About: {stock.info['longBusinessSummary']}")
```

However we have a problem, if the user asks to `List Stocks` without anything in the database it will return an error. We can fix this by checking if the len of rows in the database is equal to 0. Here is what it should look like:

```python
@match_regex(r"List Stocks")
    async def list_info_stock(self, message):
        c.execute("SELECT * FROM stocks")
        rows = c.fetchall()
        
        if len(rows) == 0:
            await message.respond('No Stocks Added, Please Use Command "Add Stock:"')
            
        for row in rows:
            stock = yf.Ticker(row[0])
            await message.respond(f"Name: {stock.info['longName']}")
            await message.respond(f"Price: {stock.info['regularMarketOpen']} {stock.info['currency']}")
            await message.respond(f"About: {stock.info['longBusinessSummary']}")
```

> Now you can add, list and find specific info about stocks. Congratulations! Goodluck with your opsdroid journey! Here is what the final code should look like:

Reviewer: ```suggestion
Now you can add, list and find specific info about stocks. Congratulations! Good luck with your opsdroid journey! Here is what the final code should look like:
```
